### PR TITLE
Components utils: use `Object.assign` instead of `{ ...spread }` syntax

### PR DESCRIPTION
**Pull request by ciampo**
## What?

The changes in this PR refactor the code in the `utils` folder of the `@wordpress/components` package, switching to `Object.assign()` instead of using the `{ ...spread }` syntax.

## Why?

While working on #39436, I discovered the following issue, which can be replicated by running `npm run build:packages` and inspecting the following generated files:

- `packages/components/build-types/utils/config-values.d.ts`: this file actually contains a couple of circular dependency errors:

<details>
<summary>See screenshot</summary>

Screenshot 2022-03-23 at 10 17 23
</details>

- `packages/components/build-types/utils/colors-values.d.ts`: in this file, we can observe that effectively the type of the `COLORS` object doesn't include the `white` key.

## How?

In both cases, I was able to solve the issue by using `Object.assign` instead of the `{ ...spread }` syntax.

## Testing Instructions

On trunk:
- `npm run distclean && npm ci && npm run build:packages`
- Inspect `packages/components/build-types/utils/config-values.d.ts` and notice the TypeScript errors.
- Inspect `packages/components/build-types/utils/colors-values.d.ts`, and notice how the type of the `COLORS` object doesn't include the `white` key.

Apply the changes from this PR:
- `npm run distclean && npm ci && npm run build:packages`
- Inspect `packages/components/build-types/utils/config-values.d.ts`, there shouldn't be any TypeScript errors
- Inspect `packages/components/build-types/utils/colors-values.d.ts`, the type of the `COLORS` object should include the `white` key.

### CODE REVIEW

File: packages/components/CHANGELOG.md [hunk 1]
```diff
@@ -4,16 +4,20 @@
...
 
 -   `BaseControl`: Convert to TypeScript ([#39468](https://github.com/WordPress/gutenberg/pull/39468)).
 
 ### New Features
 
-- Add `BorderControl` component ([#37769](https://github.com/WordPress/gutenberg/pull/37769)).
-- Add `BorderBoxControl` component ([#38876](https://github.com/WordPress/gutenberg/pull/38876)).
+-   Add `BorderControl` component ([#37769](https://github.com/WordPress/gutenberg/pull/37769)).
+-   Add `BorderBoxControl` component ([#38876](https://github.com/WordPress/gutenberg/pull/38876)).
+
```
Comment (ciampo) [inline packages/components/CHANGELOG.md:17]: Changes on these lines are only about aligning the white spacing at the beginning of every list item

File: packages/components/src/utils/config-values.js [hunk 2]
```diff
@@ -26,15 +26,16 @@ const CONTROL_PROPS = {
...
 	toggleGroupControlBackgroundColor: CONTROL_PROPS.controlBackgroundColor,
 	toggleGroupControlBorderColor: COLORS.ui.border,
 	toggleGroupControlButtonColorActive: CONTROL_PROPS.controlBackgroundColor,
 };
 
-export default {
-	...CONTROL_PROPS,
-	...TOGGLE_GROUP_CONTROL_PROPS,
+// Using Object.assign to avoid creating circular references when emitting
+// TypeScript type declarations
```
Comment (ntsekouras) [inline packages/components/src/utils/config-values.js]:
```suggestion
// TypeScript type declarations.
```

**Review by ntsekouras — APPROVED** [review]
Thanks Marco, this resolves the issue.

I'm wondering though why this happens. Is there a reference you can share to read more about it?

**Review by mirka — APPROVED** [review]
Really interesting 🤔 After [reading up](https://stackoverflow.com/questions/32925460) on it, I do see how the internal implementation could be different between spreads and assigns.  I can't seem to get a [minimal repro](https://www.typescriptlang.org/play?ts=4.4.4#code/MYewdgzgLgBAYiEMC8MDeMCGAuGByTPGAXwChRJYBlABwCcBTTAExXRgDouEQAaGYLjzAiZCtBgBBCBACWAczBsA8gCMAVg2BQOmGQrAAKNMX49+GQfhEkAlEA) so I wonder if there's something else involved too. But that's just for curiosity.

**Comment by ciampo** [comment]
I also personally don't really know why this was happening — it definitely looks it's a TypeScript bug, but I also found it difficult to reproduce it in isolation.

Somehow when using the spread syntax in the affected, TypeScript doesn't seem to be able to fully "expand" the object that is being spread. In the case of `packages/components/build-types/utils/config-values.d.ts`, TypeScript even added an `import` statement — which I don't understand the reason for 🤷 

I also found some differences between `Object.assign` and spread syntax on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals)

**Comment by ntsekouras** [comment]
I wonder if this is something an option in `tsconfig` could handle and/or if it has to do with some transpiler step 🤔

**Comment by ciampo** [comment]
> I wonder if this is something an option in `tsconfig` could handle and/or if it has to do with some transpiler step 🤔

That is also a possibility, although I haven't had the time to look into it after finding a fix. I'll put it in my list of tasks, but anyone feel free to also explore it!

**Pull request merged**